src/main/java/com/algorand/algosdk/transaction/AtomicTransactionComposer.java

```diff
@@ -0,0 +1,449 @@
+            this.txIDs = txIDs;
+            this.methodResults = methodResults;
+        }
+    }
+
+    public static class ReturnValue {
+        public String txID;
+        public byte[] rawValue;
+        public ABIValue value;
+        public Exception parseError;
+
+        public ReturnValue(String txID, byte[] rawValue, ABIValue value, Exception parseError) {
+            this.txID = txID;
+            this.rawValue = rawValue;
+            this.value = value;
+            this.parseError = parseError;
+        }
+    }
+
+    public static class TransactionSigner {
+        protected enum SignedBy {
+            BasicAccount,
+            LogicSigAccount,
+            MultiSigAccount
+        }
+
+        protected TransactionSigner.SignedBy signedBy;
+        protected Account acc;
+        protected LogicSigAccount lsa;
+        protected List<Account> msigAccounts = new ArrayList<>();
+        protected MultisigAddress msigAddr;
+
+        public TransactionSigner(Account acc) {
+            this.signedBy = SignedBy.BasicAccount;
+            this.acc = acc;
+        }
+
+        public TransactionSigner(LogicSigAccount lsa) {
+            this.signedBy = SignedBy.LogicSigAccount;
+            this.lsa = lsa;
+        }
+
+        public TransactionSigner(MultisigAddress msigAddr, byte[][] sks) throws NoSuchAlgorithmException {
+            this.signedBy = SignedBy.MultiSigAccount;
+            this.msigAddr = msigAddr;
+            for (byte[] sk : sks) this.msigAccounts.add(new Account(sk));
+        }
+
+        public SignedTransaction[] signTxnGroup(Transaction[] txnGroup, int[] indicesToSign)
+                throws NoSuchAlgorithmException, IOException {
+            List<SignedTransaction> signedTxns = new ArrayList<>();
+            for (int index : indicesToSign)
+                signedTxns.add(this.signTxn(txnGroup[index]));
+            return signedTxns.toArray(new SignedTransaction[0]);
+        }
+
+        public SignedTransaction signTxn(Transaction txn) throws NoSuchAlgorithmException, IOException {
+            if (this.signedBy == SignedBy.BasicAccount) {
+                return acc.signTransaction(txn);
+            } else if (this.signedBy == SignedBy.LogicSigAccount) {
+                return lsa.signLogicSigTransaction(txn);
+            } else if (this.signedBy == SignedBy.MultiSigAccount) {
+                List<SignedTransaction> multiStxns = new ArrayList<>();
+                for (Account acc : this.msigAccounts)
+                    multiStxns.add(acc.signMultisigTransaction(this.msigAddr, txn));
+                return Account.mergeMultisigTransactions(multiStxns.toArray(new SignedTransaction[0]));
+            } else
+                throw new IllegalArgumentException("atomic transaction signer cannot infer sign transaction by who");
+        }
+    }
+
+    public interface MethodArgument {
+    }
+
+    public static class ABIValue implements MethodArgument {
```

Comment:
I don't think there's any reason to have `MethodArgument` and `ABIValue` in this SDK. I left a similar comment on the Go SDK where I explained why: https://github.com/algorand/go-algorand-sdk/pull/258#discussion_r743308560